hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/file/CsvMetadataResolver.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.connector.file;
+
+import com.hazelcast.jet.pipeline.file.CsvFileFormat;
+import com.hazelcast.jet.pipeline.file.FileFormat;
+import com.hazelcast.jet.sql.impl.extract.CsvQueryTarget;
+import com.hazelcast.jet.sql.impl.schema.MappingField;
+import com.hazelcast.sql.impl.QueryException;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+final class CsvMetadataResolver extends MetadataResolver {
+
+    static final CsvMetadataResolver INSTANCE = new CsvMetadataResolver();
+
+    private CsvMetadataResolver() {
+    }
+
+    @Override
+    public String supportedFormat() {
+        return CsvFileFormat.FORMAT_CSV;
+    }
+
+    @Override
+    public List<MappingField> resolveAndValidateFields(List<MappingField> userFields, Map<String, ?> options) {
+        return !userFields.isEmpty() ? validateFields(userFields) : resolveFieldsFromSample(options);
+    }
+
+    private List<MappingField> validateFields(List<MappingField> userFields) {
+        for (MappingField userField : userFields) {
+            if (userField.externalName() != null) {
+                throw QueryException.error("EXTERNAL NAME not supported");
```

Comment:
The following test fails:
```java
    @Test
    public void test_fieldsSubset() {
        String name = randomName();
        sqlService.execute("CREATE MAPPING " + name + " ("
                + "\"int\" VARCHAR"
                + ", byte VARCHAR"
                + ") TYPE " + FileSqlConnector.TYPE_NAME + ' '
                + "OPTIONS ( "
                + '\'' + OPTION_FORMAT + "'='" + CSV_FORMAT + '\''
                + ", '" + FileSqlConnector.OPTION_PATH + "'='" + RESOURCES_PATH + '\''
                + ", '" + FileSqlConnector.OPTION_GLOB + "'='" + "file.csv" + '\''
                + ")"
        );

        assertRowsAnyOrder(
                "SELECT * FROM " + name,
                singletonList(new Row(
                        2147483647,
                        (byte) 127
                ))
        );
    }
```
The failure is: 
```
java.lang.AssertionError: 
Expecting:
  <[Row{[string, true]}]>
to contain exactly in any order:
  <[Row{[2147483647, 127]}]>
elements not found:
  <[Row{[2147483647, 127]}]>
and elements not expected:
  <[Row{[string, true]}]>
```

The reason is that when we specify the columns for the CSV file mapping, it doesn't extract the specified columns from the CSV file, but it simply uses the first column in the CSV under the first name etc. This is surprising behavior IMO. When I cay `create mapping t (my_field INT)`, i expect that the field `my_field` will be extracted from the CSV and not the first field.

Second, when I map a non-existing field, the query should either fail or return null for that field. I think the 2nd option is better, this will allow for schema evolution: newer files will miss some fields, but we can map them and process older and newer files together. The following test should pass too:

```java
    @Test
    public void when_mapNonExistingColumn_then_isNull() {
        String name = randomName();
        sqlService.execute("CREATE MAPPING " + name + " (non_existent_field VARCHAR) "
                + "TYPE " + FileSqlConnector.TYPE_NAME + ' '
                + "OPTIONS ( "
                + '\'' + OPTION_FORMAT + "'='" + CSV_FORMAT + '\''
                + ", '" + FileSqlConnector.OPTION_PATH + "'='" + RESOURCES_PATH + '\''
                + ", '" + FileSqlConnector.OPTION_GLOB + "'='" + "file.csv" + '\''
                + ")"
        );

        assertRowsAnyOrder("select * from " + name,
                singletonList(new Row((Object) null)));
    }
```